atc/db/resource_check_rate_limiter.go

```diff
@@ -81,44 +88,35 @@ func (limiter *ResourceCheckRateLimiter) Wait(ctx context.Context) error {
 }
 
 func (limiter *ResourceCheckRateLimiter) Limit() rate.Limit {
-	limiter.mut.Lock()
-	defer limiter.mut.Unlock()
-
 	return limiter.checkLimiter.Limit()
 }
 
 func (limiter *ResourceCheckRateLimiter) refreshCheckLimiter() error {
-	var resourceCount int
-	var resourceTypeCount int
-	err := psql.Select("COUNT(id)").
-		From("resources").
-		Where(sq.Eq{"active": true}).
+	var count int
+	err := psql.Select("COUNT(*)").
+		From("resources r").
+		Join("pipelines p ON p.id = r.pipeline_id").
+		Where(sq.Eq{
+			"r.active": true,
+			"p.paused": false,
+		}).
 		RunWith(limiter.refreshConn).
 		QueryRow().
-		Scan(&resourceCount)
-	if err != nil {
-		return err
-	}
-	err = psql.Select("COUNT(id)").
-		From("resource_types").
-		Where(sq.Eq{"active": true}).
-		RunWith(limiter.refreshConn).
-		QueryRow().
-		Scan(&resourceTypeCount)
+		Scan(&count)
 	if err != nil {
 		return err
 	}
 
-	count := resourceCount + resourceTypeCount
 	limit := rate.Limit(float64(count) / limiter.checkInterval.Seconds())
 	if count == 0 {
 		// don't bother waiting if there aren't any checkables
 		limit = rate.Inf
 	}
+	if limit < limiter.minChecksPerSecond {
+		limit = limiter.minChecksPerSecond
+	}
```

Comment:
Wild thought: can we setup a notifier to refresh the check limit every time a pipeline is saved/deleted/paused? To deal with larger installations we probably still want to rely on the periodic refresh. So it would be something like: if # of pipelines/resources < threshold, then use the notifier to refresh on pipeline changes, otherwise refresh every minute.

Haven't thought much about the complexities of that solution, happy to use the current solution if you think it's too much effort for such a small problem.

Reply:
Yeah I thought about that, but I felt like it wasn't worth the complexity it adds, especially considering the fact that if the number of resources is low then you don't really need to rate limit in the first place